File: md04/ex02b/turtle_multi_target/src/turtle_tf2_broadcaster.cpp
#include <memory>
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/transform_stamped.hpp>
#include <tf2_ros/transform_broadcaster.h>
#include <turtlesim/msg/pose.hpp>
#include <cmath>

class TurtleTf2Broadcaster : public rclcpp::Node
{
public:
  TurtleTf2Broadcaster()
  : Node("turtle_tf2_broadcaster")
  {
    tf_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);
    
    turtle1_sub_ = create_subscription<turtlesim::msg::Pose>(
      "/turtle1/pose", 10,
      std::bind(&TurtleTf2Broadcaster::turtle1_pose_cb, this, std::placeholders::_1));
      
    turtle2_sub_ = create_subscription<turtlesim::msg::Pose>(
      "/turtle2/pose", 10,
      std::bind(&TurtleTf2Broadcaster::turtle2_pose_cb, this, std::placeholders::_1));
      
    turtle3_sub_ = create_subscription<turtlesim::msg::Pose>(
      "/turtle3/pose", 10,
      std::bind(&TurtleTf2Broadcaster::turtle3_pose_cb, this, std::placeholders::_1));
  }

private:
  void turtle1_pose_cb(const turtlesim::msg::Pose::SharedPtr msg)
  {
    publish_transform(*msg, "turtle1");
  }

  void turtle2_pose_cb(const turtlesim::msg::Pose::SharedPtr msg)
  {
    publish_transform(*msg, "turtle2");
  }

  void turtle3_pose_cb(const turtlesim::msg::Pose::SharedPtr msg)
  {
    publish_transform(*msg, "turtle3");
  }

  void publish_transform(const turtlesim::msg::Pose& pose, const std::string& turtle_name)
  {
    geometry_msgs::msg::TransformStamped t;
    t.header.stamp = this->get_clock()->now();
    t.header.frame_id = "world";
    t.child_frame_id = turtle_name;
    
    t.transform.translation.x = pose.x;
    t.transform.translation.y = pose.y;
    t.transform.translation.z = 0.0;
    
    t.transform.rotation.x = 0.0;
    t.transform.rotation.y = 0.0;
    t.transform.rotation.z = sin(pose.theta / 2.0);
    t.transform.rotation.w = cos(pose.theta / 2.0);
    
    tf_broadcaster_->sendTransform(t);
  }

  std::shared_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
  rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr turtle1_sub_;
  rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr turtle2_sub_;
  rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr turtle3_sub_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<TurtleTf2Broadcaster>());
  rclcpp::shutdown();
  return 0;
}

File: md04/ex02/turtle_tf2/turtle_tf2/turtle_tf2_broadcaster.py
import rclpy
from rclpy.node import Node
from turtlesim.msg import Pose
from geometry_msgs.msg import TransformStamped
from tf2_ros import TransformBroadcaster
import math

class TurtleTfBroadcaster(Node):
    def __init__(self):
        super().__init__('turtle_tf2_broadcaster')
        self.tf_broadcaster = TransformBroadcaster(self)

        # Подписываемся на позы turtle1 и turtle2
        self.create_subscription(Pose, '/turtle1/pose', self.turtle1_pose_cb, 10)
        self.create_subscription(Pose, '/turtle2/pose', self.turtle2_pose_cb, 10)

    def turtle1_pose_cb(self, msg: Pose):
        t = TransformStamped()
        t.header.stamp = self.get_clock().now().to_msg()
        t.header.frame_id = 'world'
        t.child_frame_id = 'turtle1'
        t.transform.translation.x = msg.x
        t.transform.translation.y = msg.y
        t.transform.translation.z = 0.0
        # theta -> quaternion (z, w)
        t.transform.rotation.z = math.sin(msg.theta / 2.0)
        t.transform.rotation.w = math.cos(msg.theta / 2.0)
        self.tf_broadcaster.sendTransform(t)

    def turtle2_pose_cb(self, msg: Pose):
        t = TransformStamped()
        t.header.stamp = self.get_clock().now().to_msg()
        t.header.frame_id = 'world'
        t.child_frame_id = 'turtle2'
        t.transform.translation.x = msg.x
        t.transform.translation.y = msg.y
        t.transform.translation.z = 0.0
        t.transform.rotation.z = math.sin(msg.theta / 2.0)
        t.transform.rotation.w = math.cos(msg.theta / 2.0)
        self.tf_broadcaster.sendTransform(t)

def main(args=None):
    rclpy.init(args=args)
    node = TurtleTfBroadcaster()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
